packages/at_secondary_server/lib/src/verb/handler/pkam_verb_handler.dart

```diff
@@ -82,14 +74,99 @@ class PkamVerbHandler extends AbstractVerbHandler {
       response.errorMessage = 'pkam publickey not found';
       throw UnAuthenticatedException('pkam publickey not found');
     }
-    var isValidSignature = false;
+    var isValidSignature =
+        await _validateSignature(verbParams, sessionID, atSign, publicKey);
+    // authenticate if signature is valid
+    if (isValidSignature) {
+      atConnectionMetadata.isAuthenticated = true;
+      atConnectionMetadata.authType = pkamAuthType;
+      atConnectionMetadata.enrollmentId = enrollId;
+      response.data = 'success';
+    } else {
+      atConnectionMetadata.isAuthenticated = false;
+      response.data = 'failure';
+      logger.severe('pkam authentication failed');
+      throw UnAuthenticatedException('pkam authentication failed');
+    }
+  }
 
-    //retrieve stored secret using sessionid and atsign
-    var storedSecret = await keyStore.get('private:$sessionID$atSign');
-    storedSecret = storedSecret?.data;
+  /// Performs operations required for apkam authentication
+  /// Returns pkamPublicKey if the [enrollId] is [EnrollStatus.approved]
+  /// Otherwise returns [Response] with appropriate [Response.errorCode] and [Response.errorMessage]
+  /// Throws [KeyNotFoundException] if the key for provided [enrollId] does not exist
+  @visibleForTesting
+  Future<dynamic> handleApkamVerification(
+      String enrollId, String atSign, Response response) async {
+    String enrollmentKey =
+        '$enrollId.$newEnrollmentKeyPattern.$enrollManageNamespace$atSign';
+    var enrollData = await keyStore.get(enrollmentKey);
+    final atData = enrollData.data;
+    logger.finer('handleApkamVerification() fetched enrollData: $atData');
+    final enrollDataStoreValue =
+        EnrollDataStoreValue.fromJson(jsonDecode(atData));
+    response = verifyEnrollApproval(
+        enrollDataStoreValue.approval!.state, enrollId, response);
+    // If response.isError is true, return the response object as it contains
+    // appropriate errorCode and errorMessage that need to be displayed to the user
+    if (response.isError) {
+      return response;
+    }
+    // If response.isError is NOT true, return the public key to continue authentication
+    // using apkam public key
+    // This means that the enrollmentId provided is APPROVED
+    return enrollDataStoreValue.apkamPublicKey;
+  }
 
+  @visibleForTesting
+  Response verifyEnrollApproval(
+      String approvalState, String enrollId, Response response) {
```

Comment:
don't pass an object and modify the state inside the method.
Construct a new response object inside the method and return.

Reply:
changes made as part of [7566e7e](https://github.com/atsign-foundation/at_server/pull/1520/commits/7566e7e79435c00f0b11a050b6cbc72ebbf54487)